qiskit_optimization/problems/substitute_variables.py

```diff
@@ -0,0 +1,308 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Substitute variables of QuadraticProgram."""
+
+import logging
+from collections import defaultdict
+from math import fsum
+from typing import Dict, List, Optional, Tuple, Union
+
+from .constraint import ConstraintSense
+from .linear_expression import LinearExpression
+from .quadratic_expression import QuadraticExpression
+from .quadratic_program import QuadraticProgram
+from ..exceptions import QiskitOptimizationError
+from ..infinity import INFINITY
+
+logger = logging.getLogger(__name__)
+
+
+class SubstituteVariables:
+    """A class to substitute variables of an optimization problem with constants for other
+    variables"""
+
+    CONST = "__CONSTANT__"
+
+    def __init__(self):
+        self._src = None  # type: Optional[QuadraticProgram]
```

Comment:
I don't know if you want to consider changing these while you are doing this. Since we no longer support 3.5 we can put these as proper typehints on a the variables rather than using comments. Something perhaps we can switch over as we change code and see this.